design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,319 @@
+ 
+Regardless of which retries mechanism is applied when a timeout is configured a service may fail to deliver a result
+in time but will deliver it eventually, but the invoker (or a service mash) would've sent a duplicating request
+by that time. To avoid duplicating operations for such cases _X-Magento-Call-Id_ header
+will be added to inter-service requests to identify calls.
+It will be randomly generated. Then if a timeout reached while waiting for a service to respond the invoker
+(or a service mash) will send the same request
+again configured number of times with the same call ID. The node receiving requests will store the call ID in cache as a key and
+assign an empty value to it. When the node will have the service call result it will store it with the call ID. That way when a
+duplicate request comes in if the node can't find the call ID in cache it will execute the requested service method as usual,
+if the cache has the call ID that would mean the service method is being executed and we'll have to wait until the original process
+finishes it and writes the response to cache, if both call ID and the result are in cache then we just return the result. That way
+we will avoid executing duplicate calls when retries mechanism is applied.
+ 
+    
+#### Service mashes and balancers
+Since we going to contact services via HTTP it is possible for configured service base URLs to actually lead to service mashes or
+balancers - that way the retries and lookup mechanism may be delegated.
+
+#### MQ RPC for gateway
+Magento has another way to remotely call services - RPC via a message queue. The problem with it is that right now it can only work
+with strings as arguments for service contracts and daemons written in PHP to process queues will show slower results in handling
+multiple requests than web servers like Nginx or Apache when using RESTful web API for gateway.
+See _Magento\Framework\MessageQueue\Rpc\*_ and _Magento\Framework\MessageQueue\UseCase\RpcCommunicationsTest_ for
+more information. It is not a finished mechanism and it will take longer to adapt it for inter-service communications.
+ 
+#### gRPC for gateway
+Remote calls to services must coexist with local calls, and on the service it would make sens to use an RPC for that but
+microservices communications are more complex than just calling a class and a method on another node with provided
+arguments - we need to pass authentication information, call (request) IDs, sign requests - we need a customizable and
+extendable gateway for this.
+
+#### Front node (BFF)
+Front node is a Magento installation with Framework, *Api, *Proxy and *Webapi/*GraphQl modules.
+ 
+This is the node that will be available on a public network.
+
+#### Flow
+Entities:
+* Front node
+  * Modules installed:
+    * Magento\Framework
+    * Magento\Authorization (implementation, APIs for auth are moved to Framework)
```

Comment:
Maybe we should leave this module as is, I'm not sure that how authorization should change.